parsing/parser.mly

```diff
@@ -698,13 +698,85 @@ let mk_directive ~loc name arg =
+  match tv_type.ptyp_desc with
+  | Ptyp_var name -> mkloc name tv_type.ptyp_loc
+  | _ -> expecting_loc tv_type.ptyp_loc "type variable"
+
+(* Note [Parsing layout annotations in types]
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   We wish to be able to parse [(int list : value)] as a layout annotation
+   in a type. This is slightly fraught, though: if we see [(int : value ...]
+   as we're parsing, have we seen the beginning a layout annotation, or have we
+   seen the beginning of a function type whose argument is labeled [int]?
+
+   In order to build a non-conflicting parser with support for these layout
+   annotations, we must keep the productions for layout annotations in line
+   with those for labeled functions. The key action happens in the definition
+   of nonterminal [non_ident_atomic_type]. We have two problems to solve:
+
+   1. Before the COLON
+
+      There are two different productions for layout annotations, one beginning with
+      [LPAREN LIDENT COLON] and the other wish [LPAREN non_ident_atomic_type COLON]. The
+      former nicely overlaps with the production for labeled functions (in
+      [strict_function_type]); this overlap allows menhir to avoid conflicts.
+      We similarly must be careful to avoid conflicts between these two productions,
+      which is why we define non_ident_atomic_type, which parses types that are
+      anything other than a single identifier.
+
+   2. After the COLON
+
+      Because we need to keep the labeled-function parser and the layout-annotations
+      parser lined up until the disambiguator (either an [->] or a [)]), we still
+      must be careful after the colon. Here, we use [tuple_type], which again
+      dovetails with the productions in [strict_function_type].
+
+      Of course, a layout annotation is not quite a [tuple_type], and so
+      [check_layout_from_type] converts, reporting an error if the conversion
+      is impossible.
+
+   The current design works well, but it is not as powerful as it could be.
+   In particular, the "before the colon" part accepts only non_ident_atomic_type,
+   meaning that e.g. [(int * bool * string : value)] cannot be parsed. If we
+   want to, we can extend this treatment to tuples or even function types,
+   but doing so seems under-motivated.
+
+   [Similar story]: There is a similar overlap around [let f : ('a : value) ...].
+   Is that the beginning of, say, a function type where ['a] is the argument?
+   Or is it the beginning of a list of universal variables. To avoid conflicts
+   in this case (the non-terminal of interest is [possibly_poly], which can't
+   know which side to take), we have layout-annotation case of [typevar] parse
+   liberally, lining up with the layout-annotation case of [non_ident_atomic_type],
+   checking that the variable is really a variable and the layout annotation
+   is really a layout annotation in [check_type_var_from_type] and
+   [check_layout_from_type].
+*)
+
```

Comment:
Agreed about delicate design here. I'd be happy for something different than what I've written. We could say that labels require having no space and layouts require having a space, but that's not backward compatible and will make upstreaming harder than it would be otherwise. And it makes a spacebar significant instead of a parenthesis, which is worse. (We could require both whitespace *and* parens. Maybe that's a good idea? I don't know.)

I don't like the keywords idea: we're eventually going to allow abstract layouts (e.g. `sig layout l type t : l ... end`) and so the keywords approach will run out of steam pretty soon.

On balance, my call would be to keep what I've written. But it's more because I don't see a better option than that I like what we have, and so more brainstorming might be fruitful. Maybe add to an agenda for a type-systems meeting? (To be clear: I don't plan on adding this to an agenda myself, but would happily support you if you do.)

Reply:
(Leaving this for you to resolve so that the comment doesn't get lost.)

Reply:
Yes, I agree that if we want abstract layouts (and we can't find a clever way to syntactically distinguish them - personally I think all emojis would be good), this is an acceptable compromise.  I think it may be worth adding to the agenda to discuss with the team along with the printing stuff, in case anyone has a clever idea.

Reply:
OK. I've put this on my list of planned items for next week's type-systems meeting. Resolving, knowing we can return later. I've also made https://jira/browse/TANDC-2078 so that we don't forget this.